### PR TITLE
Runbook page

### PR DESCRIPTION
**Pull request by Cesarsar**

### CODE REVIEW

File: src/components/DownloadFileButton.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import React, {Component} from 'react';
+import {omit} from 'lodash';
+import FileSaver from 'file-saver';
+import cx from 'classnames';
+
+export default class DownloadFileButton extends Component {
```
Comment (vkramskikh) [inline src/components/DownloadFileButton.js]: I don't think this controls is needed at all - there is not so much logic. Actually it's just saveFile method which can be passed as onClick

File: src/components/runbooks/RunbookForm.js [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+import {observer} from 'mobx-react';
+import {clone} from 'lodash';
+import cx from 'classnames';
+
+import FileInput from '../FileInput';
+import DownloadFileButton from '../DownloadFileButton';
+
+@observer
+export default class RunbookForm extends Component {
+  @observable key = Date.now()
```
Comment (vkramskikh) [inline src/components/runbooks/RunbookForm.js]: I'd name it somehow more meaningful and create a separate action with meaningful name to update it. It took me about a minute to recall what this hack is for :)

File: src/components/runbooks/RunbookForm.js [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+  changeParameter = (index, attr, value) => {
+    let parameters = clone(this.props.data.parameters.slice());
```
Comment (vkramskikh) [inline src/components/runbooks/RunbookForm.js]: Why do we need both clone and slice here?

File: src/components/runbooks/RunbookForm.js [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+            onChange={onChange}
+          />
+          <p className='help-block'>{errors.name}</p>
+        </div>
+        <div className='runbook-tags'>
+          <label className='control-label'>{'Tags'}</label>
+          {data.tags.map((tag, index) => {
+            const error = (errors.tags || [])[index];
+            return (
+              <div className='row' key={index}>
```
Comment (vkramskikh) [inline src/components/runbooks/RunbookForm.js]: Can we use tag itself here? Using index for key is last resort option and here there are better options

File: src/components/runbooks/RunbookPage.js [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+      ]
+    };
+    runbookRuns.items = runbookRunsResponseBody.runs.map((run) => new RunbookRun(run));
+  }
+
+  @observable key = Date.now();
+  @observable data;
+  @observable errors = {};
+
+  @action
```
Comment (vkramskikh) [inline src/components/runbooks/RunbookPage.js]: o_O

File: src/components/runbooks/RunbookPage.js [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+  componentWillMount() {
```
Comment (vkramskikh) [inline src/components/runbooks/RunbookPage.js]: Please use constructor()

File: src/components/runbooks/RunbookPage.js [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+    const {params: {runbookId}, runbooks} = this.props;
+    this.data = new Runbook(runbooks.get(runbookId));
+  }
+
+  @action
+  onChange = (e) => {
+    const {name, value} = e.target;
+    transaction(() => {
```
Comment (vkramskikh) [inline src/components/runbooks/RunbookPage.js]: It's not necessary to use transaction withing action - all actions are transactions

**Review by vkramskikh — CHANGES_REQUESTED** [review]

File: src/components/runbooks/RunbookForm.js (hunk 4 above)
Comment (Cesarsar) [reply src/components/runbooks/RunbookForm.js]: Nope as we edit tag string here and key prop will change for the row on each user input.
Changed to 'tag'+index

File: src/components/runbooks/RunbookRunsPage.js [hunk 8]
```diff
@@ -201,19 +209,19 @@ export default class RunbookRunsPage extends Component {
                         {!regionName && <td>{runbookRun.runbook.regionId}</td>}
                         <td>
                           <Link to={
-                            '/region/' + runbookRun.regionId +
+                            '/region/' + runbookRun.runbook.regionId +
```
Comment (vkramskikh) [inline src/components/runbooks/RunbookRunsPage.js]: Please use encodeURIComponent as we do everywhere else

File: src/components/runbooks/RunbookForm.js [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+import cx from 'classnames';
+import FileSaver from 'file-saver';
+
+import FileInput from '../FileInput';
+
+@observer
+export default class RunbookForm extends Component {
+  changeParameter = (index, attr, value) => {
+    let parameters = this.props.data.parameters.slice();
+    parameters[index][attr] = value;
```
Comment (vkramskikh) [inline src/components/runbooks/RunbookForm.js]: Why do you clone parameters at all, if you're mutating the value directly anyway and it is mutated in the old array also?

**Review by vkramskikh — APPROVED** [review]

**Pull request merged**